Compliance Operator v0.1.57 release notes and enhancements added (#50778)
openshift/openshift-docs: modules/compliance-custom-node-pools.adoc

:_content-type: PROCEDURE
[id="compliance-custom-node-pools"]
= Creating custom node pools

The Compliance Operator does not maintain a copy of each node pool configuration. The Compliance Operator aggregates consistent configuration options for all nodes within a single node pool into one copy of the configuration file. The Compliance Operator then uses the configuration file for a particular node pool to evaluate rules against nodes within that pool.

> If your cluster uses custom node pools outside the default `worker` and `master` node pools, you will need to supply additional variables to ensure the Compliance Operator aggregates a configuration file for that node pool.

Reviewer: Not sure if this could be a note or could be left as is.

Author: I think I'm going to leave it as-is since we are describing the process. Might revisit that later prior to merge, though. Good looking out!

  setValues:
  - name: ocp4-var-role-master
 value: infra
 rationale: test for infra nodes
  - name: ocp4-var-role-worker
 value: infra
 rationale: test for infra nodes
  description: cis-infra-scan
----

. Add the `infra` role to the `ScanSetting` that will be stored in the `ScanSettingBinding`:
+
[source,yaml]
----
apiVersion: compliance.openshift.io/v1alpha1
kind: ScanSetting
metadata:
  name: default
  namespace: openshift-compliance
rawResultStorage:
  rotation: 3
  size: 1Gi
roles:
- worker
- master
- infra
scanTolerations:
- effect: NoSchedule
  key: node-role.kubernetes.io/master
  operator: Exists
schedule: '0 1 * * *'
----

. Create a scan using `ScanSettingBinding`:
+
[source,yaml]
----
apiVersion: compliance.openshift.io/v1alpha1
kind: ScanSettingBinding
metadata:
  name: cis
  namespace: openshift-compliance
profiles:
- apiGroup: compliance.openshift.io/v1alpha1
  kind: Profile
  name: ocp4-cis
- apiGroup: compliance.openshift.io/v1alpha1
  kind: Profile
  name: ocp4-cis-node
- apiGroup: compliance.openshift.io/v1alpha1
  kind: TailoredProfile
  name: cis-infra-tp
settingsRef:
  apiGroup: compliance.openshift.io/v1alpha1
  kind: ScanSetting
  name: default
----

> The Compliance Operator checks the runtime `KubeletConfig` through the Kubernetes `Node/Proxy` object, and then uses variables `ocp-var-role-master` and `ocp-var-role-worker` to determine the nodes it performs the check against. In the `ComplianceCheckResult`, the `KubeletConfig` rules will be shown as `ocp4-cis-kubelet-*`. The scan only passes if all selected nodes pass the check.

Reviewer: ```suggestion
The Compliance Operator checks the runtime `KubeletConfig` through the Kubernetes `Node/Proxy` object, and then uses variables such as `ocp-var-role-master` and `ocp-var-role-worker` to determine the nodes it performs the check against. In the `ComplianceCheckResult`, the `KubeletConfig` rules will be shown as `ocp4-cis-kubelet-*`. The scan only passes if all selected nodes pass this check.
```

// Module included in the following assemblies:
//
// * security/compliance_operator/compliance-operator-remediation.adoc

:_content-type: PROCEDURE
[id="compliance-custom-node-pools"]

> = Creating custom node pools

Reviewer 2: This section might be called Scanning custom node pools or similar (we're not creating them here..)

The Compliance Operator does not maintain a copy of each node pool configuration. The Compliance Operator aggregates consistent configuration options for all nodes within a single node pool into one copy of the configuration file. The Compliance Operator then uses the configuration file for a particular node pool to evaluate rules against nodes within that pool.

If your cluster uses custom node pools outside the default `worker` and `master` node pools, you will need to supply additional variables to ensure the Compliance Operator aggregates a configuration file for that node pool.

.Procedure

> . To check configuration against all pools in an example cluster containing `master`, `worker`, and `infra` node pools, set the value of `ocp-var-role-master` and `opc-var-role-worker` to `infra` in the `TailoredProfile`:

Reviewer 2: This all looks good to me but @Vincent056 might want to double check.

> . To check configuration against all pools in an example cluster containing `master`, `worker`, and `infra` node pools, set the value of `ocp-var-role-master` and `opc-var-role-worker` to `infra` in the `TailoredProfile`:

Reviewer 3: Yes, it is just an example of custom node pool

> . To check configuration against all pools in an example cluster containing `master`, `worker`, and `infra` node pools, set the value of `ocp-var-role-master` and `opc-var-role-worker` to `infra` in the `TailoredProfile`:

Reviewer 3: Maybe we should say, to scan not only master and worker node, but we scan infra node by using a tailored profile.

> . To check configuration against all pools in an example cluster containing `master`, `worker`, and `infra` node pools, set the value of `ocp-var-role-master` and `opc-var-role-worker` to `infra` in the `TailoredProfile`:

Author: Ah okay - I can change infra to generic "example" node to avoid confusion.

+
[source,yaml]
----
apiVersion: compliance.openshift.io/v1alpha1
kind: TailoredProfile
metadata:
  name: cis-infra-tp
spec:
  extends: ocp4-cis

>   title: My modified nist profile with a custom value

Reviewer 3: Maybe using `My modified NIST profile to scan infra node` is better here

>   title: My modified nist profile with a custom value

Author: As above, `My modified NIST profile to scan example node` sounds good to me.

  setValues:
  - name: ocp4-var-role-master
 value: infra
 rationale: test for infra nodes
  - name: ocp4-var-role-worker
 value: infra
 rationale: test for infra nodes
  description: cis-infra-scan
----

. Add the `infra` role to the `ScanSetting` that will be stored in the `ScanSettingBinding`:
+
[source,yaml]
----
apiVersion: compliance.openshift.io/v1alpha1
kind: ScanSetting
metadata:
  name: default
  namespace: openshift-compliance
rawResultStorage:
  rotation: 3
  size: 1Gi
roles:
- worker
- master
- infra
scanTolerations:
- effect: NoSchedule
  key: node-role.kubernetes.io/master
  operator: Exists
schedule: '0 1 * * *'
----

. Create a scan using `ScanSettingBinding`:
+
[source,yaml]
----
apiVersion: compliance.openshift.io/v1alpha1
kind: ScanSettingBinding
metadata:
  name: cis
  namespace: openshift-compliance
profiles:
- apiGroup: compliance.openshift.io/v1alpha1
  kind: Profile

>   name: ocp4-cis

Reviewer 3: `cis-infra-tp` will only scan Infra node, and `cop-cis` will scan `master` and `worker` node

>   name: ocp4-cis

Reviewer 4: What does `ocp4-cis-node` scan?

>   name: ocp4-cis

Reviewer 3: ocp4-cis-node is for cis-node profile, we have one kubelet rules that is in cis-node profile, ex. `kubelet_enable_protect_kernel_defaults`

- apiGroup: compliance.openshift.io/v1alpha1
  kind: Profile
  name: ocp4-cis-node
- apiGroup: compliance.openshift.io/v1alpha1
  kind: TailoredProfile
  name: cis-infra-tp
settingsRef:
  apiGroup: compliance.openshift.io/v1alpha1
  kind: ScanSetting
  name: default
----

The Compliance Operator checks the runtime `KubeletConfig` through the `Node/Proxy` API object, and then uses variables such as `ocp-var-role-master` and `ocp-var-role-worker` to determine the nodes it performs the check against. In the `ComplianceCheckResult`, the `KubeletConfig` rules will be shown as `ocp4-cis-kubelet-*`. The scan only passes if all selected nodes pass this check.

.Verification

> * You can verify if a rule is checked through the `Node/Proxy` object by checking if `valuesUsed` of the `ComplianceCheckResult` contains `ocp4-var-role-master` or `ocp4-var-role-worker`:

Reviewer 3: Had a second look on this, I think this is invalid, can we change this to 

All the Platform KubeletConfig rules are checked through the `Node/Proxy` object, you can find those rules by:
```
oc get rules -o json | jq '.items[] | select(.checkType == "Platform") | select(.metadata.name | contains("ocp4-kubelet-")) | .metadata.name'
```

+
[source,terminal]
----

> $ oc get ccr -n openshift-compliance -o yaml | jq '.items[] | select(.valuesUsed | contains("ocp4-var-role-master") or contains("ocp4-var-role-worker"))'`

Reviewer 3: same here as above

.Procedure

. To check the configuration against all pools in an example cluster containing `master`, `worker`, and custom `example` node pools, set the value of the `ocp-var-role-master` and `opc-var-role-worker` fields to `example` in the `TailoredProfile` object:
+
[source,yaml]
----
apiVersion: compliance.openshift.io/v1alpha1
kind: TailoredProfile
metadata:
  name: cis-example-tp
spec:
  extends: ocp4-cis
  title: My modified NIST profile to scan example nodes
  setValues:
  - name: ocp4-var-role-master
    value: example
    rationale: test for example nodes
  - name: ocp4-var-role-worker
    value: example
    rationale: test for example nodes
  description: cis-example-scan
----

. Add the `example` role to the `ScanSetting` object that will be stored in the `ScanSettingBinding` CR:
+
[source,yaml]
----
apiVersion: compliance.openshift.io/v1alpha1
kind: ScanSetting
metadata:
  name: default
  namespace: openshift-compliance
rawResultStorage:
  rotation: 3
  size: 1Gi
roles:
- worker
- master
- example
scanTolerations:
- effect: NoSchedule
  key: node-role.kubernetes.io/master
  operator: Exists
schedule: '0 1 * * *'
----

> . Create a scan using `ScanSettingBinding` CR:

Reviewer 5: > ! using noun
To avoid ambiguity, replace this gerund with either "by using" or "that uses".

+
[source,yaml]
----
apiVersion: compliance.openshift.io/v1alpha1
kind: ScanSettingBinding
metadata:
  name: cis
  namespace: openshift-compliance
profiles:
- apiGroup: compliance.openshift.io/v1alpha1
  kind: Profile
  name: ocp4-cis
- apiGroup: compliance.openshift.io/v1alpha1
  kind: Profile
  name: ocp4-cis-node
- apiGroup: compliance.openshift.io/v1alpha1
  kind: TailoredProfile
  name: cis-example-tp
settingsRef:
  apiGroup: compliance.openshift.io/v1alpha1
  kind: ScanSetting
  name: default
----

> The Compliance Operator checks the runtime `KubeletConfig` through the `Node/Proxy` API object, and then uses variables such as `ocp-var-role-master` and `ocp-var-role-worker` to determine the nodes it performs the check against. In the `ComplianceCheckResult`, the `KubeletConfig` rules will be shown as `ocp4-cis-kubelet-*`. The scan only passes if all selected nodes pass this check.

Reviewer 5: ```suggestion
The Compliance Operator checks the runtime `KubeletConfig` through the `Node/Proxy` API object and then uses variables such as `ocp-var-role-master` and `ocp-var-role-worker` to determine the nodes it performs the check against. In the `ComplianceCheckResult`, the `KubeletConfig` rules are shown as `ocp4-cis-kubelet-*`. The scan passes only if all selected nodes pass this check.
```

.Verification

* The Platform KubeletConfig rules are checked through the `Node/Proxy` object. You can find those rules by running the following command:
+
[source,terminal]
----

> $ oc get rules -o json | jq '.items[] | select(.checkType == "Platform") | select(.metadata.name | contains("ocp4-kubelet-")) | .metadata.name'

Reviewer 5: > Do not use jq in commands (unless it is truly required), because this requires users to install the jq tool.  Oftentimes, the same or similar result can be accomplished using `jsonpath` for `oc` commands.

This is out of scope for a merge review, but consider fixing it in the future.
